src/hmem_ipc_cache.c

```diff
@@ -0,0 +1,166 @@
+				entry->info.iov.iov_base, entry->info.iov.iov_len);
+		}
+	}
+	return ret;
+}
+
+static void ipc_cache_delete_region(struct ofi_mr_cache *cache, struct ofi_mr_entry *entry)
+{
+	ofi_hmem_close_handle(entry->info.iface, entry->info.ipc_mapped_addr);
+}
+
+/**
+ * @brief Open an ipc cache
+ * 
+ * @param cache[in] the ipc cache
+ * @param domain[in] the domain that the cache is attached to.
+ * @return int 0 on success, negative value otherwise.
+ */
+int ofi_ipc_cache_open(struct ofi_mr_cache **cache, struct fid_domain *domain)
+{
+	/* 
+	 * The ipc monitor should be general for any iface.
+	 * Start from FI_HMEM_CUDA and we can expand it later.
+	 */
+	struct ofi_mem_monitor *memory_monitors[OFI_HMEM_MAX] = {
+		[FI_HMEM_CUDA] = ipc_monitor,
+	};
+	struct util_domain *util_domain;
+	int ret;
+	util_domain = container_of(domain, struct util_domain, domain_fid);
+	*cache = (struct ofi_mr_cache *)calloc(1, sizeof(struct ofi_mr_cache));
+	if (!*cache)
+		return -FI_ENOMEM;
+	(*cache)->add_region = ipc_cache_add_region;
+	(*cache)->delete_region = ipc_cache_delete_region;
+	ret = ofi_mr_cache_init(util_domain, memory_monitors,
+				*cache);
+	if (ret) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"ipc cache init failed: %s\n",
+		        fi_strerror(ret));
+		free(*cache);
+		*cache = NULL;
+		return -ret;
+	}
+
+	FI_INFO(&core_prov, FI_LOG_CORE,
+		"ipc cache enabled, max_cnt: %zu max_size: %zu\n",
+		cache_params.max_cnt, cache_params.max_size);
+	return 0;
+}
+
+/**
+ * @brief Destroy the ipc cache
+ * 
+ * @param cache the ipc cache
+ */
+void ofi_ipc_cache_destroy(struct ofi_mr_cache *cache)
+{
+	ofi_mr_cache_cleanup(cache);
+	free(cache);
+	cache = NULL;
+}
+
+/**
+ * @brief Given ipc_info (with ipc_handle and the iov of the device allocation),
+ * assign the mapped_addr the mapped address of the ipc_handle.
+ * Each (ipc_handle, mapped_addr) pair is stored in ofi_mr_entry.info.ipc_info as
+ * part of each mr entry.
+ * In a cache hit, the mapped_addr is retrieved from the matched mr entry. Otherwise,
+ * the mapped_addr is obtained by opening the ipc handle.
+ * 
+ * @param[in] cache the ipc cache
+ * @param[in] ipc_info the information of the ipc to be mapped.
+ * @param[out] mapped_addr the mapped address for the ipc handle.
+ * @param[out] mr_entry the matched mr_entry of the ipc_info and mapped_addr.
+ * @return int 0 on success, negative value otherwise.
+ */
+int ofi_ipc_cache_search(struct ofi_mr_cache *cache, struct ipc_info *ipc_info,
+			  void **mapped_addr, struct ofi_mr_entry **mr_entry)
+{
+	struct ofi_mr_info info;
+	struct ofi_mr_entry *entry;
+	int ret;
+	info.iov.iov_base = (void *)ipc_info->base_address;
+	info.iov.iov_len = ipc_info->base_length;
+	info.iface = ipc_info->iface;
+	memcpy((void **)&info.ipc_handle, (void **) &ipc_info->ipc_handle, IPC_HANDLE_SIZE);
+	ret = ofi_mr_cache_search(cache, &info, &entry);
+	if (OFI_UNLIKELY(ret))
+		return ret;
+	*mapped_addr = entry->info.ipc_mapped_addr;
+	*mr_entry = entry;
```

Comment:
If we return entry, why do we need to copy ipc_mapped_addr out separately?

Reply:
That was aimed to have a separate `mapped_addr` output for this function, so it can assign the `base` in the same way with `ze_hmem_open_shared_handle` in https://github.com/ofiwg/libfabric/blob/main/prov/shm/src/smr_progress.c#L493 . I can remove it and use `entry->info.mapped_addr`